### PR TITLE
add pollution to resource annotation

### PR DESCRIPTION
**Pull request by kdada**
this makes tiller add an annotation to every kubernetes resource. Then we can find the relationship between kubernetes resources and subcharts in a release.

The annotation is like:
```
"helm.sh/pollutant":"charta/chartb/chartc"
```

@ddysher @ScorpioCPH  @superxi911 @caitong93

### CODE REVIEW

File: pkg/engine/engine.go [hunk 1]
```diff
@@ -115,6 +115,8 @@ func (e *Engine) Render(chrt *chart.Chart, values chartutil.Values) (map[string]
 type renderable struct {
 	// tpl is the current template.
 	tpl string
+	// path is the path of chart
+	path string
```
Comment (deyuanix) [inline pkg/engine/engine.go:119]: This seems orthogonal to what's described in the PR title?

File: pkg/engine/pollute.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+*/
+
+package engine
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+
+	"k8s.io/kubernetes/pkg/api"
```
Comment (deyuanix) [inline pkg/engine/pollute.go:24]: Do you need extra vendor?

File: pkg/engine/pollute.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+			panic(err)
+		}
+	}
+	announced.DefaultGroupFactoryRegistry.RegisterAndEnableAll(registered.NewOrDie(os.Getenv("KUBE_API_VERSIONS")), schema)
+	defaultSerializer = json.NewYAMLSerializer(json.DefaultMetaFactory, schema, schema)
+}
+
+// pollute adds pollutant to resource annotations. If resource is not a valid kubernetes
+// resource, it does nothing and returns original resource.
+func pollute(resource string, r *renderable) string {
```
Comment (deyuanix) [inline pkg/engine/pollute.go:77]: How's pollution used?  I guess in kubernetes-admin where you query use annotation?

**Comment by deyuanix** [comment]
Without the change (i.e. in stock helm), how do I query resources status from kubernetes?  I see `helm get`, but it won't return status, only specs.

**Comment by deyuanix** [comment]
@kdada Can u poke upstream about the issue we are facing?  Is there any way we can do instead of changing helm code, e.g. via hook?

File: pkg/engine/engine.go (hunk 1 above)
Comment (kdada) [reply pkg/engine/engine.go:119]: The value of `path` will store in the annotations of related resources.

File: pkg/engine/pollute.go (hunk 2 above)
Comment (kdada) [reply pkg/engine/pollute.go:24]: `helm` contains `k8s.io/kubernetes` package.

File: pkg/engine/pollute.go (hunk 3 above)
Comment (kdada) [reply pkg/engine/pollute.go:77]: Yeah. I query these annotations in `kubernetes-admin`.

**Comment by kdada** [comment]
@ddysher I add these annotations for identifing which subchart created these resources. `hook` can't do it.

`helm` has a cmd named `status` can query status of resources in a release. But it's not useful and detailed for us.  We need get the resources of a specified subchart.

File: pkg/engine/engine.go (hunk 1 above)
Comment (deyuanix) [reply pkg/engine/engine.go:119]: Newer version of helm has a new field `basePath`, seems it can be used to serve the purpose.  WDYT?

Comment (deyuanix) [reply pkg/engine/engine.go:119]: https://github.com/kubernetes/helm/blob/15576abd9aaf7f285a7701905dbafbb256161c87/pkg/engine/engine.go#L127

I can merge this PR anyway to unblock any ongoing work, but would like to get ur thoughts first.

Comment (kdada) [reply pkg/engine/engine.go:119]: `basePath` is a physical path of a template file like `templates/xxx.yaml` or `charts/chartX/templates/xxx.yaml`.
`path` is a logic path of a chart like `chartA/chartB/chartC`. The `path` stands for a hierarchy of subcharts.

Comment (deyuanix) [reply pkg/engine/engine.go:119]: Yes, but my understand is that if you have this path, then you can extract it? e.g. 
`charts/chartX/charts/ChartY/Charts/ChartZ/templates/xxx.yaml` --> `chartX/chartY/chartZ`

Not a big deal though, will merge if we agree on this.

Comment (kdada) [reply pkg/engine/engine.go:119]: Yes, we can extract a logic path from `basePath`. We can do it in next `rebase`.

**Comment by deyuanix** [comment]
Can u open a PR first to setup travis CI.

**Comment by deyuanix** [comment]
@kdada what's the status?

**Comment by kdada** [comment]
@ddysher 
I have ran circleci and pass the test. But it only can set my repo `kdada/helm`.  I can't add `caicloud/helm` in the circleci.

**Comment by ddysher** [comment]
Comment

**Comment by deyuanix** [comment]
@kdada You may want to open a new PR or send new commit to trigger CI.

**Comment by kdada** [comment]
@ddysher  Update test and skip deployment

**Comment by deyuanix** [comment]
:rocket:
